### ISSUE TITLE
feature: compile source code transformation to binary code transformation

### ISSUE THREAD

**Issue by monperrus** [INRIA/spoon#1834]
For the programmer, it is much easier to write a transformation at the source code level, because she is familiar with the language constructs. However, for applicability, it is better to be able to apply the transformations at the binary level. There are different options, incl: 1) compiling a source code transformation in Spoon to a binary code transformation in ASM/Javassist 2) applying transformation after decompilation.

Super fun task, in Spoon's roadmap for a long time.

Are you ready to take on the challenge?

**Comment by monperrus** [comment]
This tool compute diff on bytecode 
https://github.com/scala/jardiff

May be useful for this task.

Thanks @nharrand

**Comment by surli** [comment]
I'm a bit puzzled by the usecase: if a user wants to write a transformation at the source code level, then he must know the source code level, so this one must be available? Or do you imagine the case of analyzing the code using a decompiler to write a transfo and then apply it to the jar? 
 
Now I don't know much the field but I'm a bit afraid about the optimization realized by compilers.
A naive example would be a for-loop changed as a while: then the user would write a transfo to catch that loop, except that it doesn't exist anymore in the binaries. Do you think this kind of case could occur?

**Comment by monperrus** [comment]
> Or do you imagine the case of analyzing the code using a decompiler to write a transfo and then
> apply it to the jar?
Exactly! @nharrand is making great progress on the decompilation aspect of it.

**Comment by monperrus** [comment]
Having AST generators #1960 would be super useful here

**Comment by monperrus** [comment]
@andrewbwogi some related work:
[MetaWSL and meta-transformations in the fermaT transformation system](http://www.gkc.org.uk/martin/papers/metatrans-t.pdf) (2005)
